### DISCUSSION TITLE
Jie's code posted

### DISCUSSION THREAD

**Discussion by JNing0** [QuickPay-Operational-Performance/Data-and-code#81] [General]
@vibhuti6 My code is [here](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/codes/Rcode_nQrtrNew_1107.R). Let me know if you have any questions!

BTW, where can I find the latest data set?

**Comment by vibhuti6** [comment]
Thanks, Jie. This issue has the link to the latest data: https://github.com/QuickPay-Operational-Performance/Data-and-code/issues/79

**Reply by vibhuti6** [reply]
Could you share a toy example of how you are calculating bi-quarterly delays? 

If I understand correctly, in your code, the data frame "data.nQ[[1]]" has quarterly delays and "data.nQ[[2]]" has delays measured over two quarters. 

I checked the values in these for project ID "CONT_AWD_0001_9700_DACA8797D0011_9700", and the bi-quarterly ones don't seem correct. 

Here's the screenshot:
![Screen Shot 2021-11-16 at 6 26 50 PM](https://user-images.githubusercontent.com/47764977/142082613-87dc053f-e7ad-46f2-8586-fa2fe9f4ae00.png)

Shouldn't the delay corresponding to "2010-06-30" in the first table be zero?

Also, by truncating the data, we have removed the observation corresponding to "2011-03-31" which was there in the original dataset:
![Screen Shot 2021-11-16 at 6 35 05 PM](https://user-images.githubusercontent.com/47764977/142083096-8982fd6a-0011-49fb-97d7-f4ca1d0626d1.png)

**Reply by JNing0** [reply]
I changed the definition a bit... Shall we have a quick zoom call so that I can walk you through the code?

**Reply by vibhuti6** [reply]
I won’t be able to talk right now unfortunately. Could we set up some time for Thursday instead? I can talk at any time except 1-2pm EST on Thu. Please let me know.

In the meantime, it would be helpful if you could please explain the rationale for counting 61 days twice in the above example.

Also, I compared our results for quarterly delay. That’s happening mainly due to difference in approaches towards winsorization/truncation, so is easy to reconcile:
- My results used the non-zero winsorized delays where 5% winsorization was done on the full sample. This gives the treatment effect of around 15 days.
- Yours were on the data truncated at 2.5% after removing non-zero delays. However, even in that, the treatment effect would be around 20 days when you consider only treat*post (so no multiple quarters).
- The effect is around 25 days when you winsorize at 2.5% after removing non-zero delays. 

Here’s the [code](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/codes/edits_jie_code.R) for reference, built on yours.

**Reply by JNing0** [reply]
I just checked the code and saw a bug in aggregating the quarters which I fixed. But after rerunning the model, the results remain pretty much the same. Here is the result on the example project after fixing the code. Thanks for checking!

![image](https://user-images.githubusercontent.com/58864118/142114318-95be4a90-4fe8-49b7-bd56-c64963147957.png)

A few things in aggregating the quarterly delays: 
- In my definition, the bi-quarterly delay of 2010-6-30 is the total delay in the 2nd quarter and the _following_ quarter of 2010. The annual delay of 2010-6-30 is the total delay in the 2nd quarter in 2010 and the _following_ three quarters. 
- We should have the same number of time points before and after QP when running the DiD. 
- The launch of QP, in the 2nd quarter of 2011, should not fall in the middle of an interval. That's why I started a new bi-quarter internal and a new year at the 2nd quarter.

The quarterly results are not affected by the bug. So yes, without the time-dependent treatment effect, the time*post term is lower, about 20 days. This is consistent with the time-dependent result in Table 2 we saw this afternoon (see [here](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/QP1/Time-dependent%20treatment%20effects.pdf)) as there is a significant dip two quarters after QuickPay. 

But over multiple quarters, the results stay the same, about 20 days, after fixing the code and rerunning the model. The treatment effect remains fairly stable regardless of the aggregation level. I don't see it doubling...

I have updated the code on GitHub. Let me know if you still would like to meet.

**Reply by vibhuti6** [reply]
Not sure I fully understand this definition. Let's work through an example. 

- Here's the original data for project ID "CONT_AWD_0001_9700_DCA20002D5003_9700":
![Screen Shot 2021-11-16 at 11 46 27 PM](https://user-images.githubusercontent.com/47764977/142137463-26ce31b7-770b-4704-aa52-60fba4ac5f4b.png)

- These are the bi-quarterly delays with your definition. What happens to the observations in last three rows with your definition, and why?
![Screen Shot 2021-11-16 at 11 47 03 PM](https://user-images.githubusercontent.com/47764977/142137553-037bed24-a665-4653-81ed-6cc3cf69739f.png)

- This is how I calculated bi-quarterly delay. I take six-month intervals starting June 2010. Specifically, delay for June 2010 semester would be "quarterly delay in Mar 2010" + "quarterly delay in June 2010"
![Screen Shot 2021-11-16 at 11 55 16 PM](https://user-images.githubusercontent.com/47764977/142137600-135b348b-eab3-4f5a-a71b-0d015d4256fe.png)

**Reply by JNing0** [reply]
As I said earlier, I dropped the first (Qrtr 1) and last quarter (Qrtr 10) so that we have the same number of time periods before and after QP.
So for the bi-quarterly delay, there are four bi-quarter intervals (two before QP and two after):
- Qrtr 2-3, 2010
- Qrtr 4, 2010 - Qrtr 1, 2011
- Qrtr 2-3, 2011
- Qrtr 4, 2011 - Qrtr 1, 2012

For the annual delay, there are the two annual intervals:
- Qrtr 2, 2010 - Qrtr 1, 2011
- Qrtr 2, 2011 - Qrtr 1, 2012

Note that both the bi-quarterly and annual aggregation are done within one year before and after QP.

In the example you gave, the last row is dropped as it is quarter 10 in the data. The sum of the second and third of the last rows is zero, which is also dropped in data.lm, which contains only nonzero delays. If you look at data.nQ, which contains the raw aggregated delay after truncating at 2.5% level, you should see a zero row.

**Reply by vibhuti6** [reply]
Thanks. Looks like the different intervals in our models give different results. I will try to run my code with different intervals as well to confirm.

One other thing I noticed is that when you aggregate after truncating the data, some of the observations that shouldn't be there are retained. For instance, consider project ID "CONT_AWD_0001_9700_DACA8797D0011_9700" again. Here's the original data and the bi-quarterly delay with your code:

![Screen Shot 2021-11-17 at 11 29 16 AM](https://user-images.githubusercontent.com/47764977/142241511-32155275-0156-4f34-89c6-d71e6b22a579.png)

The delay corresponding to 2010-12-31 should be (61+4042)=4103 days with your definition. Ideally, this observation should be removed from the data, but instead we are counting its delay as 61 days. This is unlikely to affect the results too much, but could you check once by truncating *after* aggregation?

**Reply by JNing0** [reply]
Good question. I truncated the nonzero delay data _before_ aggregation. In other words, I took the original quarterly delay, took out the nonzero delays, and threw out the upper and lower 2.5%. Aggregation came after that so was done on _truncated_ quarterly delays. The reason is that if the delay in a particular quarter is unreliable, then we should only take that out, not the reliable ones. So no further truncation is done on aggregated delays.

Coming back to your example, the data entry with delay 4042 was deleted but the entry with delay 61 remains. So after aggregation, we are left with 61.

It is a good idea to try different intervals, just make sure that
- we have the same time points before and after QP
- QP does not happen in the middle of a multi-quarter interval.

Let me know if you have other questions!

**Reply by vibhuti6** [reply]
I understand that. My concern is that we are underestimating/miscalculating the delay over the full semester by picking only the quarter with 61 days and discarding the one with 4021 days. The entire semester observation should be removed for a consistent comparison. Could you please compare the results when you truncate after aggregating?

QuickPay is still happening in the middle of the interval as it was launched in April 2011. But yes, the gap is not as extreme as in other case. Also some of the difference in our results is coming from the winsorization approach, as in the case of one quarter. I am re-running my code with different approaches, and will post the updated results.

**Reply by vibhuti6** [reply]
So I think the difference in our two-quarter delay results was due to winsorization approach and not different intervals. 
Specifically, winsorization replaces extreme values. In the full sample, most of the delays are zero so the lower replacement threshold was zero. That is, all negative delays get replaced by zero. So when we consider the subsample of non-zero winsorized delay, it effectively meant we are excluding negative delay observations.

I re-ran the full sample and non-zero sample regressions with different truncation/winsorization approaches. In essence, the results are what we would expect:
- Treatment effect doubles with full sample (that includes zero delays).
- Treatment effect does not change when you consider non-zero delays.

I also truncated after aggregation, and there's no change as expected but I think this is a more accurate measure. 

Here are the latest [results](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/qp_first_implementation/qp_first_pc_delay_two_quarters.pdf) with your definition of two-quarter delay, and the corresponding [code](https://github.com/QuickPay-Operational-Performance/Data-and-code/blob/master/qp_first_implementation/qp_first_pc_delay_two_quarters.Rmd).

**Reply by JNing0** [reply]
Great. Thank you for checking!

**Comment by JNing0** [comment]
I forgot to say, I did run the logistic regressions with clustering at the project level (as you can see from the code), and the estimated errors are smaller than under the PSC level. So to be conservative about our estimates, I'd personally go with the PSC-level clustering.

For linear regressions on the nonzero delays, we should NOT have clustering on the project level, as 75% projects with nonzero delays delay only once. This means that we are not really clustering. The PSC level generates larger estimate errors than, e.g., NAICS level clustering, so I used the PSC level clustering.
